### PR TITLE
Core/adapting space to communicator

### PR DESCRIPTION
**Pull request by RiccardoRossi**
**Description**
This is a Baby step towards being able to eventually use our own matrix implementation through the kratos.
on one side this adds communicator in the contstruction of matrices, something that is crucial in mpi and does no harm in serial, on the other hand it removes some resizing methods which cannot be ported to mpi.
Please note that resize by pointer is left, however i would vouch for also removing it.

Please mark the PR with appropriate tags: 
-

**Changelog**
Adding DataCommunicator to the Space interface 
Removing Resize from the interface of the space

### CODE REVIEW

File: applications/StructuralMechanicsApplication/custom_python/structural_mechanics_python_application.cpp [hunk 1]
```diff
@@ -189,6 +186,9 @@ PYBIND11_MODULE(KratosStructuralMechanicsApplication,m)
     /* Reset equations ids "flag" */
     KRATOS_REGISTER_IN_PYTHON_VARIABLE(m,RESET_EQUATION_IDS);
 
+    /* Reset displacement "flag" needed for GeoMechanicalApplication*/
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_python/structural_mechanics_python_application.cpp]: Why the variables are changed?

File: applications/TrilinosApplication/trilinos_space.h [hunk 2]
```diff
@@ -302,16 +302,16 @@ class TrilinosSpace
         rX.PutScalar(A);
     }
 
-    static void Resize(MatrixType& rA, SizeType m, SizeType n)
-    {
-        KRATOS_THROW_ERROR(std::logic_error, "Resize is not defined for Trilinos Sparse Matrix", "")
+    // static void Resize(MatrixType& rA, SizeType m, SizeType n)
```
Comment (loumalouomega) [inline applications/TrilinosApplication/trilinos_space.h]: Remove?

File: applications/TrilinosApplication/trilinos_space.h [hunk 3]
```diff
@@ -115,17 +115,17 @@ class TrilinosSpace
...
     }
 
-    static VectorPointerType CreateEmptyVectorPointer()
+    static VectorPointerType CreateEmptyVectorPointer(const DataCommunicator& rComm)
     {
         return VectorPointerType(nullptr);
     }
 
-    static MatrixPointerType CreateEmptyMatrixPointer(Epetra_MpiComm& Comm)
+    static MatrixPointerType CreateEmptyMatrixPointer(Epetra_MpiComm& Comm) //TODO: unify with the previous ones
```
Comment (loumalouomega) [inline applications/TrilinosApplication/trilinos_space.h:128]: Comment should be in a Doxygen block?

File: kratos/python/add_strategies_to_python.cpp [hunk 4]
```diff
@@ -130,17 +130,18 @@ namespace Kratos
...
-        {
-            dummy.Resize(A, i1, i2);
-        }
-
-        template< typename TSpaceType >
-        void ResizeVector(TSpaceType& dummy, typename TSpaceType::VectorType& x, unsigned int i1)
-        {
-            dummy.Resize(x, i1);
-        }
+//         template< typename TSpaceType >
```
Comment (loumalouomega) [inline kratos/python/add_strategies_to_python.cpp:135]: Remove

File: kratos/python/add_strategies_to_python.cpp [hunk 5]
```diff
@@ -221,8 +222,8 @@ namespace Kratos
 
             binder.def("ClearMatrix", ClearMatrix<TSpaceType>);
             binder.def("ClearVector", ClearVector<TSpaceType>);
-            binder.def("ResizeMatrix", ResizeMatrix<TSpaceType>);
-            binder.def("ResizeVector", ResizeVector<TSpaceType>);
+            // binder.def("ResizeMatrix", ResizeMatrix<TSpaceType>);
```
Comment (loumalouomega) [inline kratos/python/add_strategies_to_python.cpp:218]: Remove?

File: kratos/solving_strategies/builder_and_solvers/native_csr_block_builder_and_solver.h [hunk 6]
```diff
@@ -0,0 +1,1660 @@
...
+        ) override
+    {
+        KRATOS_TRY
+
+        if (rModelPart.MasterSlaveConstraints().size() != 0) {
+            BuildMasterSlaveConstraints(rModelPart);
+
+            KRATOS_ERROR << " MPCs not yet implemented" << std::endl;
+
+            // // We compute the transposed matrix of the global relation matrix
```
Comment (loumalouomega) [inline kratos/solving_strategies/builder_and_solvers/native_csr_block_builder_and_solver.h]: Clean up?

File: kratos/spaces/kratos_space.h [hunk 7]
```diff
@@ -365,36 +365,34 @@ class KratosSpace
         rX.SetValue(A);
     }
 
-    static void Resize(MatrixType& rA, SizeType m, SizeType n)
-    {
-        rA.resize(m, n, false);
-    }
+    // static void Resize(MatrixType& rA, SizeType m, SizeType n)
```
Comment (loumalouomega) [inline kratos/spaces/kratos_space.h]: Clean up?

File: kratos/spaces/ublas_space.h [hunk 8]
```diff
@@ -584,46 +574,46 @@ class UblasSpace
...
-#ifndef _OPENMP
-        std::fill(rA.begin(), rA.end(), TDataType());
-#else
-        DataType* vals = rA.value_data().begin();
-        #pragma omp parallel for firstprivate(m)
-        for(int i=0; i<static_cast<int>(m); ++i)
-            vals[i] = TDataType();
-#endif
-    }
+//     template<class TOtherMatrixType>
```
Comment (loumalouomega) [inline kratos/spaces/ublas_space.h]: Clean up?

**Review by loumalouomega — CHANGES_REQUESTED** [review]
Comments

File: applications/StructuralMechanicsApplication/custom_python/structural_mechanics_python_application.cpp (hunk 1 above)
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/custom_python/structural_mechanics_python_application.cpp]: not needed i am trying to revert this but having problems

File: applications/TrilinosApplication/trilinos_space.h (hunk 3 above)
Comment (RiccardoRossi) [reply applications/TrilinosApplication/trilinos_space.h:128]: no, this is "internal suggestion" should not be in docs

File: applications/TrilinosApplication/trilinos_space.h (hunk 2 above)
Comment (RiccardoRossi) [reply applications/TrilinosApplication/trilinos_space.h]: ok

File: kratos/python/add_strategies_to_python.cpp (hunk 5 above)
Comment (RiccardoRossi) [reply kratos/python/add_strategies_to_python.cpp:218]: i would prefer to keep it for a while, to optionally throw a meaningful error

**Comment by philbucher** [comment]
@RiccardoRossi I am very positive about the changes in the Space
I have a bunch of comments for the new B&S, could you please move it to a separate PR so that the space changes can go in already?

**Review by philbucher — DISMISSED** [review]
Awesome 👏

**Review by philbucher — DISMISSED** [review]

**Comment by philbucher** [comment]
seems like the master merge went wrong somehow

**Comment by RiccardoRossi** [comment]
in essence this one is outdated, since we would like to eventually (soonish) get rid of the space in the new redesign of the strategies and stuff

**Pull request closed without merging**